File: Dynamic Programming/JavaScript Algorithms/BootCamp/Problem Algorithms/productOfThree.js
/*1: Create a function that, given a list of integers, returns the
highest product between three of those numbers. For example,
given the list [1, 10, 2, 6, 5, 3] the highest product
would be 10 * 6 * 5 = 300
*/
//Naive Solution Using sort() method 
const highestProduct=(array)=>{
	//const sortDesc=array.sort((a, b)=> b - a);
	const array2=array.sort((a, b)=> a - b);
	const n=array2.length
	const a=array2[0]*array2[1]*array2[n-1]
	const b=array2[n-1]*array2[n-2]*array2[n-3]
	// return Math.max(a, b)
	return a > b ? a : b
}

// Time Complexity ===>  O(nlogn) because of sort() built-in method
//Space Complexity ===> O(n)

//Using Single scan Iteration method
//Works for array length >= 5
const highestProduct2=(array)=>{
	let min1=array[0];
	let min2=array[0];
	let max1=array[0];
	let max2=array[0];
	let max3=array[0];

	for (let num of array){
		if (num <= min1){
			min2 = min1;
			min1 = num
		} else if (num <= min2){// if num lies between min1 and min2
			min2 = num
		}
		if (num >= max1){
			max3 = max2;
			max2 = max1;
			max1 = num;
		}
		else if (num >= max2 && num < max1){// if num lies between max1 and max2
			max3 = max2;
			max2 = num;
		}
		else if (num >= max3 && num < max2){
			max3 = num
		}
	}
	return Math.max(max1 * max2 * max3, min1 * min2 * max1)
}
// Time Complexity ===>  O(n)
//Space Complexity ===> O(1)

console.log(highestProduct([1, 10, 2, 6, 5, 3, -12, -8]))//960
console.log(highestProduct([1, 10, 2, 6, 5, 3]))//300
console.log(highestProduct([1, 10, 2, 6, 5, 3, 12, -8]))//720
console.log(highestProduct2([1, 10, 2, 6, 5, 3, -12, -8]))//960
console.log(highestProduct2([1, 10, 2, 6, 5, 3]))//300
console.log(highestProduct2([1, 10, 2, 6, 5, 3, 12, -8]))//720

File: Dynamic Programming/Python Algorithms/productOfThree.py
# 1: Create a function that, given a list of integers, returns the
# highest product between three of those numbers. For example,
# given the list [1, 10, 2, 6, 5, -12, -8] the highest product
# would be 10 * 6 * 5 = 300

#Naive Solution Using sorted()
def highest_product(list):
	#list2=sorted(list, reverse=True) Even increases Time Complexity further to O(n * n) ~ O(n^2) Quadratic
	list2=sorted(list)
	a=list2[0]*list2[1]*list2[-1]
	b=list2[-1]*list2[-2]*list2[-3]
	print(a, b)
	return max(a, b)

# Time Complexity ==> O(nlogn)
# Space Complexity ==> O(n)

#Better Solution Using Single Scan Approach
def highest_product2(list):
	min1 = list[0] 
	min2 = list[0] 
	max1 = list[0] 
	max2 = list[0] 
	max3 = list[0]
	for num in list:
		if (num > max1):
			max3 = max2
			max2 = max1
			max1 = num
		elif (num > max2 and num < max1):
			max3 = max2
			max2 = num
		elif( num > max3 and num < max2):
			max3 = num

		if (num < min1):
			min2 = min1
			min1 = num
		elif (num < min2 and num > min1):
			min2 = num
	return max(min1 * min2 * max1, max1 * max2 * max3)

# Time Complexity ==> O(n) Linear
# Space Complexity ==> O(1)

my_list= [1, 10, 2, 6, 5, 3, -12, -8]
print(highest_product(my_list))
print(highest_product2(my_list))



